File: server/controllers/shareController.js
const Document = require("../models/Document");
const User = require("../models/user");

// 📩 Share document by email (owner only)


// URL: POST /api/share/share

//After Enter Email and Click on Invite Button This method is called
exports.shareDocumentByEmail = async (req, res) => {
  try {
    //documentId=> Which document to share
    //email=> Whom to share (Input)
    const { documentId, email } = req.body;
    const senderId = req.user._id;

    const receiver = await User.findOne({ email });
    if (!receiver) return res.status(404).json({ message: "User not found" });

    // Owner must exist on the note

    //Document Owner Must Be Sender
    const note = await Document.findOne({ _id: documentId, user: senderId });
    if (!note) return res.status(404).json({ message: "Document not found or unauthorized" });

    if (receiver._id.toString() === senderId.toString()) {
      return res.status(400).json({ message: "Cannot invite yourself" });
    }

    // Avoid duplicates
    const already =
      note.collaborators.some(
        (c) => c.user.toString() === receiver._id.toString()
      ) ||
      note.pendingCollaborators?.some(
        (p) => p.user.toString() === receiver._id.toString()
      );

    if (already)
      return res.status(400).json({ message: "Already shared with this user" });


    // Add to pending collaborators
    note.pendingCollaborators = note.pendingCollaborators || [];
    note.pendingCollaborators.push({ user: receiver._id, status: "pending" });

    note.isShared = true;
    note.modifiedBy = senderId;
    await note.save();

    res.json({ message: "Invitation sent successfully" });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};


// 📜 Get pending invites for logged-in user
// URL: GET /api/share/invites

//This method is called to get all pending invites for logged in user
exports.getPendingInvites = async (req, res) => {
  try {
    //Logged in user Who want to see pending invites
    const userId = req.user._id;

    //Check all documents where logged in user is in pendingCollaborators
    const invites = await Document.find({ "pendingCollaborators.user": userId })
      .populate("user", "name email")
      .populate("pendingCollaborators.user", "name email")
      .select("title user pendingCollaborators");

    res.json(invites);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// ✅ Accept invite

//This method is called when user accept invite And click on Accept Button

//Work of this method is Remove user from pendingCollaborators and add to collaborators array
exports.acceptInvite = async (req, res) => {
  try {
    //Get documentId from request body
    const { documentId } = req.body;
    const userId = req.user._id;

    //Find document by Id
    const note = await Document.findById(documentId);
    if (!note) return res.status(404).json({ message: "Document not found" });

    // Only invited users can accept

    //Check in pendingCollaborators array Of note Variable(note => get from request body After click on Accept Button of panding invites)
    const isPending = note.pendingCollaborators?.some((p) => p.user.toString() === userId.toString());
    if (!isPending) return res.status(400).json({ message: "No pending invite found" });

    // Remove from pending and add to collaborators
    note.pendingCollaborators = (note.pendingCollaborators || []).filter(
      (p) => p.user.toString() !== userId.toString()
    );

    // Avoid duplicate collaborators
    const alreadyCollab = note.collaborators.some((c) => c.user.toString() === userId.toString());
    if (!alreadyCollab) note.collaborators.push({ user: userId, role: "editor" });
    note.isShared = true;
    note.modifiedBy = userId;
    await note.save();

    res.json({ message: "Invitation accepted" });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// ❌ Reject invite

//This method is called when user reject invite And click on Reject Button
//Work of this method is Remove user from pendingCollaborators array
exports.rejectInvite = async (req, res) => {
  try {
    const { documentId } = req.body;
    const userId = req.user._id;

    const note = await Document.findById(documentId);
    if (!note) return res.status(404).json({ message: "Document not found" });

    const before = (note.pendingCollaborators || []).length;
    note.pendingCollaborators = (note.pendingCollaborators || []).filter(
      (p) => p.user.toString() !== userId.toString()
    );
    const after = note.pendingCollaborators.length;
    if (before === after) return res.status(400).json({ message: "No pending invite found" });

    note.modifiedBy = userId;
    await note.save();

    res.json({ message: "Invitation declined" });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

File: server-spring/src/main/java/com/docsmaster/server/controllers/ShareController.java
package com.docsmaster.server.controllers;

import com.docsmaster.server.dto.DocumentMapper;
import com.docsmaster.server.models.Collaborator;
import com.docsmaster.server.models.Document;
import com.docsmaster.server.models.PendingCollaborator;
import com.docsmaster.server.models.User;
import com.docsmaster.server.repositories.CollaboratorRepository;
import com.docsmaster.server.repositories.DocumentRepository;
import com.docsmaster.server.repositories.PendingCollaboratorRepository;
import com.docsmaster.server.repositories.UserRepository;
import com.docsmaster.server.security.services.UserDetailsImpl;
import com.docsmaster.server.services.SocketService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/share")
@RequiredArgsConstructor
public class ShareController {

    private final DocumentRepository documentRepository;
    private final PendingCollaboratorRepository pendingCollaboratorRepository;
    private final CollaboratorRepository collaboratorRepository;
    private final UserRepository userRepository;
    private final SocketService socketService;

    @PostMapping("/share")
    public ResponseEntity<?> shareDocumentByEmail(@AuthenticationPrincipal UserDetailsImpl userDetails, @RequestBody Map<String, String> payload) {
        String documentId = payload.get("documentId");
        String email = payload.get("email");
        String senderId = userDetails.getId();
        
        Optional<Document> docOpt = documentRepository.findById(documentId);
        if (docOpt.isEmpty()) {
            return ResponseEntity.notFound().build();
        }
        Document document = docOpt.get();
        
        // Owner must exist on the note
        if (!document.getUser().getId().equals(senderId)) {
            return ResponseEntity.status(404).body(Map.of("message", "Document not found or unauthorized"));
        }
        
        Optional<User> receiverOpt = userRepository.findByEmail(email);
        if (receiverOpt.isEmpty()) {
            return ResponseEntity.status(404).body(Map.of("message", "User not found"));
        }
        User receiver = receiverOpt.get();
        
        if (receiver.getId().equals(senderId)) {
            return ResponseEntity.badRequest().body(Map.of("message", "Cannot invite yourself"));
        }
        
        // Avoid duplicates
        boolean alreadyCollab = document.getCollaborators().stream()
                .anyMatch(c -> c.getUser().getId().equals(receiver.getId()));
        boolean alreadyPending = pendingCollaboratorRepository.findByDocumentIdAndEmail(documentId, email).isPresent();
        
        if (alreadyCollab || alreadyPending) {
            return ResponseEntity.badRequest().body(Map.of("message", "Already shared with this user"));
        }

        PendingCollaborator pending = PendingCollaborator.builder()
                .document(document)
                .email(email)
                .role("editor")
                .invitedAt(LocalDateTime.now())
                .build();
        document.getPendingCollaborators().add(pending);
        document.setIsShared(true);
        document.setModifiedBy(userRepository.findById(senderId).orElse(null));
        documentRepository.save(document);
        
        return ResponseEntity.ok(Map.of("message", "Invitation sent successfully"));
    }

    @GetMapping("/invites")
    public ResponseEntity<?> getPendingInvites(@AuthenticationPrincipal UserDetailsImpl userDetails) {
        User user = userRepository.findById(userDetails.getId()).orElseThrow();
        List<PendingCollaborator> invites = pendingCollaboratorRepository.findByEmail(user.getEmail());
        
        var inviteList = invites.stream().map(invite -> Map.of(
            "_id", invite.getId(),
            "document", Map.of(
                "_id", invite.getDocument().getId(),
                "title", invite.getDocument().getTitle()
            ),
            "role", invite.getRole(),
            "invitedAt", invite.getInvitedAt()
        )).toList();
        
        return ResponseEntity.ok(inviteList);
    }

    @PostMapping("/accept")
    public ResponseEntity<?> acceptInvite(@AuthenticationPrincipal UserDetailsImpl userDetails, @RequestBody Map<String, String> payload) {
        String documentId = payload.get("documentId");
        User user = userRepository.findById(userDetails.getId()).orElseThrow();
        
        Optional<PendingCollaborator> pendingOpt = pendingCollaboratorRepository.findByDocumentIdAndEmail(documentId, user.getEmail());
        if (pendingOpt.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("message", "No invitation found"));
        }
        PendingCollaborator pending = pendingOpt.get();
        
        Document document = pending.getDocument();
        
        // Remove pending
        document.getPendingCollaborators().remove(pending);
        
        // Add collaborator if not already present
        boolean alreadyCollab = document.getCollaborators().stream()
                .anyMatch(c -> c.getUser().getId().equals(user.getId()));
                
        if (!alreadyCollab) {
            Collaborator collaborator = Collaborator.builder()
                    .document(document)
                    .user(user)
                    .role("editor")
                    .addedAt(LocalDateTime.now())
                    .build();
            document.getCollaborators().add(collaborator);
        }
        
        document.setIsShared(true);
        document.setModifiedBy(user);
        documentRepository.save(document);
        
        com.docsmaster.server.dto.DocumentDto dto = DocumentMapper.toDto(document);
        socketService.sendToRoom(documentId, "participantsChanged", Map.of("documentId", documentId, "collaborators", dto.getCollaborators()));
        return ResponseEntity.ok(Map.of("message", "Invitation accepted"));
    }

    @PostMapping("/reject")
    public ResponseEntity<?> rejectInvite(@AuthenticationPrincipal UserDetailsImpl userDetails, @RequestBody Map<String, String> payload) {
        String documentId = payload.get("documentId");
        User user = userRepository.findById(userDetails.getId()).orElseThrow();
        
        Optional<PendingCollaborator> pendingOpt = pendingCollaboratorRepository.findByDocumentIdAndEmail(documentId, user.getEmail());
        if (pendingOpt.isPresent()) {
            pendingCollaboratorRepository.delete(pendingOpt.get());
        }
        
        return ResponseEntity.ok(Map.of("message", "Invitation rejected"));
    }
}
